pkg/vault/contracts/ProtocolFeePercentagesProvider.sol

```diff
@@ -0,0 +1,149 @@
+     * @param protocolYieldFee The protocol yield fee
+     * @param areFactoryFeesSet Flag indicating fees have been set (allows zero values)
+     */
+    struct FactoryProtocolFees {
+        uint64 protocolSwapFeePercentage;
+        uint64 protocolYieldFeePercentage;
+        bool areFactoryFeesSet;
+    }
+
+    IBalancerContractRegistry private immutable _trustedContractRegistry;
+    IProtocolFeeController private immutable _protocolFeeController;
+
+    uint256 private immutable _maxProtocolSwapFeePercentage;
+    uint256 private immutable _maxProtocolYieldFeePercentage;
+
+    // Factory address => FactoryProtocolFees
+    mapping(IBasePoolFactory => FactoryProtocolFees) private _factoryDefaultFeePercentages;
+
+    constructor(
+        IVault vault,
+        IProtocolFeeController protocolFeeController,
+        IBalancerContractRegistry trustedContractRegistry
+    ) SingletonAuthentication(vault) {
+        _protocolFeeController = protocolFeeController;
+        _trustedContractRegistry = trustedContractRegistry;
+
+        if (protocolFeeController.vault() != vault) {
+            revert WrongProtocolFeeControllerDeployment();
+        }
+
+        // These values are constant in the `ProtocolFeeController`.
+        (_maxProtocolSwapFeePercentage, _maxProtocolYieldFeePercentage) = protocolFeeController
+            .getMaximumProtocolFeePercentages();
+    }
+
+    /// @inheritdoc IProtocolFeePercentagesProvider
+    function getProtocolFeeController() external view returns (IProtocolFeeController) {
+        return _protocolFeeController;
+    }
+
+    /// @inheritdoc IProtocolFeePercentagesProvider
+    function getFactorySpecificProtocolFeePercentages(
+        address factory
+    ) external view returns (uint256 protocolSwapFeePercentage, uint256 protocolYieldFeePercentage) {
+        FactoryProtocolFees memory factoryFees = _getValidatedProtocolFees(factory);
+
+        protocolSwapFeePercentage = factoryFees.protocolSwapFeePercentage;
+        protocolYieldFeePercentage = factoryFees.protocolYieldFeePercentage;
+    }
+
+    /// @inheritdoc IProtocolFeePercentagesProvider
+    function setFactorySpecificProtocolFeePercentages(
+        address factory,
+        uint256 protocolSwapFeePercentage,
+        uint256 protocolYieldFeePercentage
+    ) external authenticate {
+        // Validate the fee percentages; don't store values that the `ProtocolFeeCollector` will reject.
+        if (protocolSwapFeePercentage > _maxProtocolSwapFeePercentage) {
+            revert IProtocolFeeController.ProtocolSwapFeePercentageTooHigh();
+        }
+
+        if (protocolYieldFeePercentage > _maxProtocolYieldFeePercentage) {
+            revert IProtocolFeeController.ProtocolYieldFeePercentageTooHigh();
+        }
+
+        // Ensure precision checks will pass.
+        _protocolFeeController.ensureValidPrecision(protocolSwapFeePercentage);
+        _protocolFeeController.ensureValidPrecision(protocolYieldFeePercentage);
```

Comment:
The precision is fixed in the vault, so we can also replicate the logic in this contract. We don't necessarily need to rely on the fee controller for that now that the deployed version doesn't really have this public function.